stability-tests/daa/daa_test.go

```diff
@@ -0,0 +1,340 @@
+			// takes a few hundred nanoseconds, so we spin in place instead of e.g. calling time.Sleep()
+			for {
+				targetHashNanoseconds := targetHashNanosecondsFunction(time.Since(startTime))
+				hashElapsedTimeNanoseconds := time.Since(hashStartTime).Nanoseconds()
+				if hashElapsedTimeNanoseconds >= targetHashNanoseconds {
+					break
+				}
+			}
+
+			hashDuration := time.Since(hashStartTime)
+			hashDurations = append(hashDurations, hashDuration)
+			if len(hashDurations) > averageHashRateSampleSize {
+				hashDurations = hashDurations[1:]
+			}
+
+			if *isFinished {
+				return
+			}
+		}
+		miningDuration := time.Since(miningStartTime)
+		miningDurations = append(miningDurations, miningDuration)
+		if len(miningDurations) > averageBlockRateSampleSize {
+			miningDurations = miningDurations[1:]
+		}
+
+		if *isFinished {
+			return
+		}
+
+		averageMiningDuration := calculateAverageDuration(miningDurations)
+		averageHashNanoseconds := calculateAverageDuration(hashDurations).Nanoseconds()
+		averageHashesPerSecond := hashNanosecondsToHashesPerSecond(averageHashNanoseconds)
+		blockDAGInfoResponse, err := rpcClient.GetBlockDAGInfo()
+		if err != nil {
+			t.Fatalf("GetBlockDAGInfo: %s", err)
+		}
+		difficultyDelta := blockDAGInfoResponse.Difficulty - previousDifficulty
+		previousDifficulty = blockDAGInfoResponse.Difficulty
+		blocksMined++
+		t.Logf("Mined block. Took: %s, average block mining duration: %s, "+
+			"average hashes per second: %d, difficulty delta: %f, time elapsed: %s, blocks mined: %d",
+			miningDuration, averageMiningDuration, averageHashesPerSecond, difficultyDelta, time.Since(startTime), blocksMined)
+
+		if *isFinished {
+			return
+		}
+
+		_, err = rpcClient.SubmitBlock(templateBlock)
+		if err != nil {
+			t.Fatalf("SubmitBlock: %s", err)
+		}
+	})
+
+	averageBlocksPerSecond := calculateAverageDuration(miningDurations).Seconds()
+	expectedAverageBlocksPerSecond := float64(1)
+	deviation := math.Abs(expectedAverageBlocksPerSecond - averageBlocksPerSecond)
+	if deviation > blockRateDeviationThreshold {
+		t.Errorf("Block rate deviation %f is higher than threshold %f. Want: %f, got: %f",
+			deviation, blockRateDeviationThreshold, expectedAverageBlocksPerSecond, averageBlocksPerSecond)
+	}
+}
+
+func hashNanosecondsToHashesPerSecond(hashNanoseconds int64) int64 {
+	return time.Second.Nanoseconds() / hashNanoseconds
+}
+
+func runForDuration(duration time.Duration, runFunction func(isFinished *bool)) {
+	isFinished := false
+	go func() {
+		for !isFinished {
+			runFunction(&isFinished)
+		}
+	}()
+	time.Sleep(duration)
+	isFinished = true
+}
+
+func runKaspad(t *testing.T) func() {
+	appDir, err := common.TempDir("kaspad-daa-test")
+	if err != nil {
+		t.Fatalf("TempDir: %s", err)
+	}
+
+	kaspadRunCommand, err := common.StartCmd("KASPAD",
+		"kaspad",
+		common.NetworkCliArgumentFromNetParams(&dagconfig.DevnetParams),
+		"--appdir", appDir,
+		"--logdir", appDir,
```

Comment:
No need to specify logdir if you want it inside appDir.
What current config will do is a weird structure where you have 
```
appDir/
--- > kaspa-devnet
--- > --- > data
--- > kaspad.log
```